# In-Context & Few-shot Learning with Prompts

### Import required packages & environment variables

In [224]:
import os
import json
import openai
from dotenv import load_dotenv, find_dotenv
from IPython.display import HTML
_ = load_dotenv(find_dotenv())
openai.api_key = os.environ['OPENAI_API_KEY']
from langchain_openai import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from langchain_core.prompts.few_shot import FewShotPromptTemplate
from langchain.schema.output_parser import StrOutputParser

### Data Preparation
Let's prepare 15 train instances for the few shot learning & 10 test instances for validation.

In [225]:
with open('CONVFINQA/data/train.json') as f:
    train_data = json.load(f)

train_list = []
for entry in train_data[:15]:
    table = entry['annotation']['amt_table']
    question_sequence = entry['annotation']['dialogue_break']
    turn_program = entry['annotation']['turn_program']
    
    follow_up_questions = ''
    
    for i in range(len(question_sequence)-1):
        follow_up_questions += f'Intermediate Question: {question_sequence[i]}\n'
        follow_up_questions += f'Intermediate Solution/Turn Program: {turn_program[i]}\n'

    train_list.append({
        'context': table,
        'questions': follow_up_questions,
        'last_question': question_sequence[-1],
        'answer': turn_program[-1]
    })

In [226]:
with open('CONVFINQA/data/test_private.json') as f:
    test_data = json.load(f)

def generate_wiki_table(table_data):
    # Initialize the wiki table string
    wiki_table = "<table class='wikitable'>"

    # Iterate over each row in the table data
    for row in table_data:
        wiki_table += "<tr>"
        # Iterate over each cell in the row
        for cell in row:
            wiki_table += f"<td>{cell}</td>"
        wiki_table += "</tr>"

    # Close the table tag
    wiki_table += "</table>"

    return wiki_table

test_list = []
for entry in test_data[:10]:
    table = generate_wiki_table(entry['table'])
    question_sequence = entry['annotation']['dialogue_break']

    follow_up_questions = ''
    
    for i in range(len(question_sequence)-1):
        follow_up_questions += f'Intermediate Question: {question_sequence[i]}\n'

    last_question = question_sequence[-1]

    test_list.append({
        'context': table,
        'questions': follow_up_questions,
        'last_question': last_question
    })


### Prompt to Try the Base Model

In [227]:
prompt_template = PromptTemplate(
    input_variables=["context", "questions", "last_question"],
    template = "Instruction: I am a highly intelligent bot. I can have conversations with the user to answer a series of questions.\
Later questions may depend on previous questions to answer. You need to provide me with the series of questions as\
the context and I will answer the last question with a multi-step mathematical solution which is a turn program. We use symbols, such as #0,\
#1. Output gives the turn program or steps to calculate the final value. Prompt format:context: {context} questions: {questions} last_question: {last_question}"
)

train_instance = train_list[0]


prompt_template.format(
    context= train_instance['context'],
    questions= train_instance['questions'],
    last_question= train_instance['last_question']
    )
model = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0.0)
output_parser = StrOutputParser()
chain = prompt_template | model | output_parser
chain.invoke({'context':train_instance['context'],
            'questions': train_instance['questions'],
            'last_question': train_instance['last_question']})

"context: <table class='wikitable'><tr><td>1</td><td>2008</td><td>year ended june 30 2009 2008</td><td>year ended june 30 2009 2008</td><td>year ended june 30 2009</td></tr><tr><td>2</td><td>net income</td><td>$ 103102</td><td>$ 104222</td><td>$ 104681</td></tr><tr><td>3</td><td>non-cash expenses</td><td>74397</td><td>70420</td><td>56348</td></tr><tr><td>4</td><td>change in receivables</td><td>21214</td><td>-2913 ( 2913 )</td><td>-28853 ( 28853 )</td></tr><tr><td>5</td><td>change in deferred revenue</td><td>21943</td><td>5100</td><td>24576</td></tr><tr><td>6</td><td>change in other assets and liabilities</td><td>-14068 ( 14068 )</td><td>4172</td><td>17495</td></tr><tr><td>7</td><td>net cash from operating activities</td><td>$ 206588</td><td>$ 181001</td><td>$ 174247</td></tr></table>\nquestions: Intermediate Question: what is the net cash from operating activities in 2009?\nIntermediate Solution/Turn Program: 206588\nIntermediate Question: what about in 2008?\nIntermediate Solution/Tur

### Few shot prompt & In-Context Learning

In [257]:
def few_shot_learning(few_shot_count, test_number):
    few_shot_examples = train_list[:few_shot_count]
    

    example_prompt = PromptTemplate(
        input_variables=["context", "questions", "last_question", "answer"],
        template="context: {context} questions: {questions}last_question: {last_question} \n{answer}"
    )

    few_shot_prompt = FewShotPromptTemplate(
        examples=few_shot_examples,
        example_prompt=example_prompt,
        suffix="Instruction: I am a highly intelligent bot. I can have conversations with the user to answer a series of questions.\
Later questions may depend on previous questions to answer. You need to provide me with the series of questions as\
the context and I will answer the last question with a multi-step mathematical solution which is a turn program. We use symbols, such as #0,\
#1. Output gives the turn program or steps to calculate the final value. Prompt format:context: {context} questions: {questions} last_question: {last_question}",
        input_variables=["context", "questions", "last_question"],
    )

    test_instance = train_list[test_number]

    #print(few_shot_prompt.format(context=test_instance['context'],
    #                             questions=test_instance['questions'],
    #                             last_question=test_instance['last_question']))

    chain = few_shot_prompt | model | output_parser
    print(chain.invoke({'context':test_instance['context'],
                'questions': test_instance['questions'],
                'last_question': test_instance['last_question']}))


### Generate answer for the last question in 6th instance in train data file

In [262]:
test_instance = train_list[5]
display(HTML(test_instance['context']))


In [264]:
few_shot_learning(0, 5)
few_shot_learning(1, 5)
few_shot_learning(2, 5)
few_shot_learning(3, 5)
few_shot_learning(4, 5)
few_shot_learning(5, 5)

last_solution: divide(5923147, add(2530454, 5923147))
subtract(2530454, 5923147), divide(#0, 5923147)
add(2530454, 5923147), divide(#0, 5923147)
add(2530454, 5923147), divide(#0, 5923147)
add(2530454, 5923147), divide(#2, 5923147)
add(2530454, 5923147), divide(#2, 5923147)


### Explanation
With zero-shot learning the model take a correct attempt to answer the last question.

When there is only one instance for few-shot learning the model generates an incorrect answer.

When only two or three few-shot examples are given, the model may rely heavily on those examples to generate the output, leading to a correct result. In this case the model, might have learned from the additional examples to produce the correct output.

In last two cases, despite the correct answer being 'add(2530454, 5923147), divide(#0, 5923147)' based on the input and few-shot examples, the model generated 'add(2530454, 5923147), divide(#2, 5923147)', which does not accurately reflect the provided context. This discrepancy suggests that the model may have 'hallucinated' or generated incorrect information based on its internal processing rather than the actual input.